react-js/react-part4-full-apps.md

```diff
@@ -0,0 +1,343 @@
+
+## Guided Practice
+
+Let's setup a small app.
+
+- `npx create-react-app todo-cra`
+- _grab a coffee while it installs everything_
+- `cd todo-cra`
+- `yarn start`
+
+The app should open in a new browser tab automatically and show a spinning atom logo.
+
+- `code .` to open the generated project in your editor
+- Find `App.js` in the `src` folder
+- Change the contents of `<p>Hello Techtonica!</p>` and save
+- The browser should live update automatically 🥳
+
+### Folder Structure
+
+Review the [existing files](https://create-react-app.dev/docs/folder-structure) of the project.
+
+The most important files are
+
+1. `src/App.css` - all the styles for your React components
+1. `public/index.html` - the non-React content of your web pages, e.g. favicon, static footer, etc
+1. `src/App.js` - mounts your root component
+
+#### `App.css`
+
+Change the color of the "Learn React" link to `#b36ff6` or another color you like on https://colors.lol/. Changes to CSS will live update as well.
+
+#### `index.html`
+
+- Change the page title to "TODO"
+- Find a nice favicon on https://www.flaticon.com/
+- Make this the new favicon
+
+Now let's add the main application logic to `App.js`. We'll keep everything in one file for now but as your app grows, you probably want to split components to their own files.
+
+### Starter Components
+
+Create a new file.
+
+```jsx
+// Todo.js
+import React from 'react';
+
+class Todo extends React.Component {
+  render() {
+    const { todo } = this.props;
+    return <div>{todo.text}</div>;
+  }
+}
+export default Todo;
+```
+
+```jsx
+// App.js
+import React from 'react';
+import Todo from './Todo';
+import './App.css';
+
+class Todo extends React.Component {
+  render() {
+    const { todo } = this.props;
+    return <div>{todo.text}</div>;
+  }
+}
+
+class App extends React.Component {
+  state = {
+    todos: [{ text: 'Walk dog' }, { text: 'Feed cat' }]
+  };
+  render() {
+    const { todos } = this.state;
+    return (
+      <div className="App">
+        <h1>Todos</h1>
+        <div>
+          {todos.length &&
+            todos.map((todo, idx) => <Todo key={idx} todo={todo} />)}
+        </div>
+      </div>
+    );
+  }
+}
+
+export default App;
+```
+
+You should see the heading and a single todo.
+
+#### Adding isCompleted
+
+Since this is a task app we want to mark tasks as complete. So let's change the shape of the todo data to include an `isCompleted` boolean.
+
+```js
+state = {
+  todos: [
+    { text: 'Walk dog', isCompleted: false },
+    { text: 'Feed cat', isCompleted: false }
+  ]
+};
+```
+
+Now let's update the `<Todo/>` component to render `isCompleted` as a checkbox. For attributes see [MDN Checkbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox).
+
+```jsx
+class Todo extends React.Component {
+  render() {
+    const {
+      todo: { text, isCompleted }
```

Comment:
Do they know enough about destructuring for this?

Reply:
Ooo fair. Looks like maybe no. Simplifying to not be more than one level deep

https://github.com/Techtonica/curriculum/blob/main/javascript/destructuring.md